src/envoy/http/service_control/handler_impl_test.cc

```diff
@@ -421,11 +428,43 @@ TEST_F(HandlerTest, HandlerCheckNotNeeded) {
   EXPECT_CALL(mock_check_done_callback_, onCheckDone(Status::OK));
   handler.callCheck(headers, *mock_span_, mock_check_done_callback_);
 
-  // no api key is set on this info
+  // No api key in request, and check is not called.
+  ReportRequestInfo expected_report_info;
+  initExpectedReportInfo(expected_report_info);
+  expected_report_info.status = Status::OK;
+  expected_report_info.operation_name = "get_no_key";
+  expected_report_info.api_key_state = ApiKeyState::NOT_CHECKED;
+
+  EXPECT_CALL(*mock_call_,
+              callReport(MatchesReportInfo(expected_report_info, headers,
+                                           response_headers, resp_trailer_)));
+  handler.callReport(&headers, &response_headers, &resp_trailer_);
+}
+
+TEST_F(HandlerTest, HandlerCheckNotNeededWithUntrustedApiKey) {
+  // Test: If the operation does not require check, check should return OK
+  utils::setStringFilterState(*mock_stream_info_.filter_state_,
+                              utils::kFilterStateOperation, "get_no_key");
+  TestRequestHeaderMapImpl headers{
+      {":method", "GET"}, {":path", "/echo"}, {"x-api-key", "invalid-key"}};
+  TestResponseHeaderMapImpl response_headers{
+      {"content-type", "application/grpc"}};
+  ServiceControlHandlerImpl handler(headers, mock_stream_info_, "test-uuid",
+                                    *cfg_parser_, test_time_, stats_);
+
+  EXPECT_CALL(*mock_call_, callCheck(_, _, _)).Times(0);
+  EXPECT_CALL(*mock_call_, callQuota(_, _)).Times(0);
+  EXPECT_CALL(mock_check_done_callback_, onCheckDone(Status::OK));
```

Comment:
Can we add a case where Check get network_failure

Reply:
Done in ClientCache tests